setup.ps1

```diff
@@ -0,0 +1,287 @@
+  if ($PythonCmd) {
+    $VersionOutput = & $PythonCmd.Source --version 2>&1
+    if ($LASTEXITCODE -eq 0) {
+      $Version = $VersionOutput | Select-String -Pattern "Python (\d+\.\d+\.\d+)" | ForEach-Object { $_.Matches.Groups[1].Value }
+      Write-Log "Python version $Version found using executable '$PythonExecutable'."
+      return $true
+    }
+    else {
+      Write-Log "Python executable '$PythonExecutable' not working correctly." -Level 'ERROR'
+      return $false
+    }
+  }
+  else {
+    Write-Log "Python executable '$PythonExecutable' not found." -Level 'ERROR'
+    return $false
+  }
+}
+
+function Find-PythonExecutable {
+  $PythonExecutables = @("python", "python3.12", "python3.11", "python3.10", "python3.9", "python3", "C:\Users\$env:USERNAME\AppData\Local\Programs\Python\Python312\python.exe", "C:\Users\$env:USERNAME\AppData\Local\Programs\Python\Python311\python.exe", "C:\Users\$env:USERNAME\AppData\Local\Programs\Python\Python310\python.exe", "C:\Users\$env:USERNAME\AppData\Local\Programs\Python\Python39\python.exe", "C:\Python311\python.exe", "C:\Python310\python.exe", "C:\Python39\python.exe")
+
+  
+  foreach ($Executable in $PythonExecutables) {
+    if (Test-PythonExecutable -PythonExecutable $Executable) {
+      return $Executable
+    }
+  }
+
+  return $null
+}
+function Main {
+  "Starting the operations..." | Out-File $LogFilePath -Append
+  "Current directory: $(Get-Location)" | Out-File $LogFilePath -Append
+
+  # Exit on lower versions than Python 3.9 or when Python executable not found
+  $PythonExecutable = Find-PythonExecutable
+  if ($null -eq $PythonExecutable) {
+    Write-Host "Error: No suitable Python executable found. Please ensure that Python 3.9 or higher is installed and available in the system PATH."
+    Exit 1
+  }
+
+  # Define the path to the Python executable in the virtual environment
+  $ActivateVenv = "$VenvDir\Scripts\Activate.ps1"
+
+  # Check if the virtual environment exists, if not, create it
+  if (-Not (Test-Path $ActivateVenv)) {
+    Write-Log "Virtual environment not found. Creating virtual environment..." -Level 'ERROR'
+    & $PythonExecutable -m venv $VenvName 2>&1 | Out-File $LogFilePath -Append
+    if ($LASTEXITCODE -ne 0) {
+      Write-Log "Failed to create virtual environment." -Level 'ERROR'
+      Exit-Script -exitCode 1
+    }
+    else {
+      Write-Log "Virtual environment created."
+    }
+  }
+
+  # Activate the virtual environment and check if it was successful
+  Write-Log "Virtual environment found. Activating virtual environment..."
+  & $ActivateVenv 2>&1 | Out-File $LogFilePath -Append
+  # Check if virtual environment is activated
+  if ($env:VIRTUAL_ENV) {
+    Write-Host "Virtual environment is activated at: $($env:VIRTUAL_ENV)"
+  }
+  else {
+    Write-Log "Failed to activate virtual environment." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+  
+  # Ensure pip
+  python -m ensurepip --default-pip 2>&1 | Out-File $LogFilePath -Append
```

Comment:
I'm not entirely sure if this is necessary.

We're in the virtual environment at this point ... and even the documentation says 

> In most cases, end users of Python shouldn’t need to invoke this module directly

[source](https://docs.python.org/3/library/ensurepip.html)

Reply:
I don't know mate, I really needed it. Otherwise, I got pip module is not installed when I did python -m pip on the venv python.exe or even when I invoked pip.exe in the venv folder directly.

Reply:
well that works for me 👍